Sources/PackageGraph/PinsStore.swift

```diff
@@ -133,28 +134,54 @@ extension PinsStore: JSONSerializable {
 extension PinsStore.Pin: JSONMappable, JSONSerializable {
     /// Create an instance from JSON data.
     public init(json: JSON) throws {
-        let name: String? = json.get("package")
-        let url: String = try json.get("repositoryURL")
-        let identity = PackageIdentity(url: url)
-        let ref = PackageReference(identity: identity, path: url)
-        self.packageRef = name.flatMap(ref.with(newName:)) ?? ref
+        // backwards compatibility 12/2020
```

Comment:
What's the experience when an older tools version tries to open the pin version?  It doesn't look as if the pins file is versioned... we could probably add something to 5.4 to provide a better error message, but that would of course not help older builds.  (I'm assuming that the identity changes are not planned for 5.4...)

Reply:
There should be a version field in there, it might be added at another place. I think the JSON we are creating here is wrapped under "pins" in the final file.

IIRC, the experience with different versions of resolved files is pretty poor. The initial version of target-based dependencies changed the format and we rolled that back because of the poor experience.

Reply:
`SimplePersistence` is what adds the "version" key.

Reply:
Thanks!  Good to know that there's a version.

So that (the poor experience causing a rollback) sounds ominous.  Is there anything that could/should be fixed in 5.4 so that it doesn't happen when this lands in `main`?

Reply:
@abertelrud @neonichu the change Is backwards compatible (can read older version) but not forward compatible (older tools won't be able to read the newer format). do you think we need to make it forward compatible, or is gracefully ignoring (emitting a warning) the pin file in such case good enough?

Reply:
No, forward compatibility isn't needed.  I was just asking about the user experience of using an old toolchain with a new `Package.resolved`.  I didn't see a version number,  but it turns out that there is one that's added by other code.  So the only remaining question is whether separately we should do something in 5.4 to improve the bad experience that @neonichu mentioned (even just a warning to indicate that the resolved file is being ignored since it's using an unsupported version, or something like that).  That could be a different PR.